File: cp4/cs/geppetto.py
from operator import inv
import sys

# Read input
inputs = sys.stdin.readlines()
n, m = map(int, inputs[0].split())
invalid = set()
for i in range(m):
    a, b = map(int, inputs[i + 1].split())
    if a > b:
        a, b = b, a
    invalid.add((a - 1, b - 1))

ways = 0
for mask in range(1 << n):
    add_ways = True
    for a, b in invalid:
        if mask & (1 << a) and mask & (1 << b):
            add_ways = False
            break
    ways += add_ways

print(ways)

File: cp4/cs/geppetto.cpp
#include <bits/stdc++.h>
using namespace std;

int n, m, a, b;
unordered_set<int> invalid;

int main() {
    // Fast I/O
    ios_base::sync_with_stdio(0);
    cin.tie(NULL);

    // Read input
    cin >> n >> m;
    for (int i = 0; i < m; i++) {
        cin >> a >> b;
        // 0-indexed
        a--, b--;
        if (a > b) swap(a, b);
        int invMask = (1 << a) | (1 << b);
        invalid.insert(invMask);
    }

    // Try all possible combinations in bitmask representation
    int ways = 0;
    for (int mask = 0; mask < (1 << n); mask++) {
        int can_work = 1;
        // Check if the current mask doesn't contain any invalid pair
        for (int invMask : invalid) {
            // Count number of 1-bits
            int num_bits = __builtin_popcount(mask & invMask);
            if (num_bits == 2) {
                can_work = 0;
                break;  // Early termination
            }
        }
        ways += can_work;
    }

    // Output answer
    cout << ways << endl;
}